chainio/clients/avsregistry/writer.go

```diff
@@ -2,86 +2,225 @@ package avsregistry
+	}
+	operatorStateRetriever, err := opstateretriever.NewContractOperatorStateRetriever(
+		operatorStateRetrieverAddr,
+		ethClient,
+	)
+	if err != nil {
+		return nil, err
+	}
+	serviceManagerAddr, err := registryCoordinator.ServiceManager(&bind.CallOpts{})
+	if err != nil {
+		return nil, err
+	}
+	blsApkRegistryAddr, err := registryCoordinator.BlsApkRegistry(&bind.CallOpts{})
+	if err != nil {
+		return nil, err
+	}
+	blsApkRegistry, err := blsapkregistry.NewContractBLSApkRegistry(blsApkRegistryAddr, ethClient)
+	if err != nil {
+		return nil, err
+	}
+	stakeRegistryAddr, err := registryCoordinator.StakeRegistry(&bind.CallOpts{})
+	if err != nil {
+		return nil, err
+	}
+	stakeRegistry, err := stakeregistry.NewContractStakeRegistry(stakeRegistryAddr, ethClient)
+	if err != nil {
+		return nil, err
+	}
+	delegationManagerAddr, err := stakeRegistry.Delegation(&bind.CallOpts{})
+	if err != nil {
+		return nil, err
+	}
+	elReader, err := elcontracts.BuildELChainReader(delegationManagerAddr, ethClient, logger)
+	if err != nil {
+		return nil, err
+	}
+	return NewAvsRegistryChainWriter(
+		serviceManagerAddr,
+		registryCoordinator,
+		operatorStateRetriever,
+		stakeRegistry,
+		blsApkRegistry,
+		elReader,
+		logger,
+		ethClient,
+		txMgr,
+	)
+}
+
+// TODO(samlaf): clean up this function
 func (w *AvsRegistryChainWriter) RegisterOperatorWithAVSRegistryCoordinator(
 	ctx context.Context,
+	operatorEcdsaPrivateKey *ecdsa.PrivateKey,
+	operatorToAvsRegistrationSigSalt [32]byte,
+	operatorToAvsRegistrationSigExpiry *big.Int,
+	blsKeyPair *bls.KeyPair,
 	quorumNumbers []byte,
-	pubkey blsregistrycoordinator.BN254G1Point,
 	socket string,
-) (*types.Receipt, error) {
+) (*gethtypes.Receipt, error) {
 	w.logger.Info("registering operator with the AVS's registry coordinator")
+	// params to register bls pubkey with bls apk registry
+	operatorAddr := crypto.PubkeyToAddress(operatorEcdsaPrivateKey.PublicKey)
+	g1HashedMsgToSign, err := w.registryCoordinator.PubkeyRegistrationMessageHash(&bind.CallOpts{}, operatorAddr)
+	if err != nil {
+		return nil, err
+	}
+	signedMsg := utils.ConvertToBN254G1Point(
+		blsKeyPair.SignHashedToCurveMessage(utils.ConvertBn254GethToGnark(g1HashedMsgToSign)).G1Point,
+	)
+	G1pubkeyBN254 := utils.ConvertToBN254G1Point(blsKeyPair.GetPubKeyG1())
+	G2pubkeyBN254 := utils.ConvertToBN254G2Point(blsKeyPair.GetPubKeyG2())
+	pubkeyRegParams := regcoord.IBLSApkRegistryPubkeyRegistrationParams{
+		PubkeyRegistrationSignature: signedMsg,
+		PubkeyG1:                    G1pubkeyBN254,
+		PubkeyG2:                    G2pubkeyBN254,
+	}
+
+	// params to register operator in delegation manager's operator-avs mapping
+	msgToSign, err := w.elReader.CalculateDelegationApprovalDigestHash(
+		&bind.CallOpts{}, operatorAddr, w.serviceManagerAddr, operatorToAvsRegistrationSigSalt, operatorToAvsRegistrationSigExpiry)
+	if err != nil {
+		return nil, err
+	}
+	operatorSignature, err := crypto.Sign(msgToSign[:], operatorEcdsaPrivateKey)
```

Comment:
we should just pass signerFn and sign?

Reply:
hmm I see. you need to sign specific msg - butyou should be able to create `types.Transaction` type out of this right?

Reply:
out of what? there's no tx here, we're just signing a msg and passing it as argument to the registerOperator function. See my response to your previous comment.

Reply:
responding to above. closing this.